### DISCUSSION TITLE
How can I achieve space-between layout by Avalonia?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#11932] [Q&A] [answered]
I want to achieve space-between layout that just like `justify-content: space-between;` in css(https://www.w3schools.com/cssref/tryit.php?filename=trycss3_justify-content2).
![image](https://github.com/AvaloniaUI/Avalonia/assets/30720036/403a4320-978a-4363-a8b3-6e9d128e0847)

What I want to do is distribute items in layout evenly. The first item is flush with the start, the last is flush with the end

PS: The number of items in the layout is indeterminate

How can I achieve this?

**Comment by jp2masa** [comment, marked as answer]
I tried to implement flexbox for Avalonia some time ago, it mostly works: [nuget](https://www.nuget.org/packages/jp2masa.Avalonia.Flexbox), [github](https://github.com/jp2masa/Avalonia.Flexbox). I didn't update it to rc1.1 yet, but if you need it I can release a new version to nuget.

**Reply by kyuranger** [reply]
I always thought that the official control could solve this problem. If you can, please update your Nuget Package. I think many people will need to use this.

**Reply by maxkatz6** [reply]
@jp2masa I think control like this can be added to the https://github.com/AvaloniaUI/Avalonia.Labs, if you wish.
We were thinking about adding experimental panels to the Labs repo (flex? adaptive grid? autolayout?), possibly moving them to the core package in the future.

**Reply by kyuranger** [reply]
This is a great idea.

**Reply by jp2masa** [reply]
@kyuranger I updated the repo just now, nuget should be available soon if you still need it.

@maxkatz6 sounds good, I'll continue the discussion in AvaloniaUI/Avalonia.Labs#8.

**Comment by maxkatz6** [comment]
@kyuranger in general, you can achieve columns layout with either Grid or UniformGrid, where the last one is more suitable for dynamic lists. After that, you can apply margins/alignments on items.

**Reply by kyuranger** [reply]
But uniform grid has to set its columns while number of items in the layout is indeterminate in my project.

**Reply by maxkatz6** [reply]
UniformGrid can contain an indeterminate number of items. It will add rows if items are overflowing.

If you always need to have a single row with columns divided by the items, you can bind `<UniformGrid Columns="{Binding MyCollection.Count}" />` or similar. With a converter, you can define a custom logic depending on the input (number of items or UniformGrid.Bounds.Width, or both with multi converter).

**Reply by kyuranger** [reply]
Thanks a lot.

**Reply by timunie** [reply]
I've learned recently that setting `Rows=1` and leaving columns unset will make this effect without any special binding.

**Reply by kyuranger** [reply]
I have just tested the UniformGrid, it cannot achieve the first item is flush with the start, the last is flush with the end.

**Reply by maxkatz6** [reply]
You can add positive margins on each item and negative margin on whole uniform grid.

**Reply by kyuranger** [reply]
All right.

**Reply by mozesa** [reply]
> You can add positive margins on each item and negative margin on whole uniform grid.

It's awesome, plus so much simpler than
```xaml
<ItemsControl.Styles>
    <Style Selector="ItemsControl > ContentPresenter:nth-child(n+2)">
        <Setter Property="Margin" Value="10 0 0 0" />
    </Style>
</ItemsControl.Styles>
```

**Reply by dbriard** [reply]
You can also use ItemsRepeater control with UniformGridLayout which have a Items justification property that do what you want.

**Reply by kyuranger** [reply]
11 RC1 seems has not ItemsRepeater any more.

**Reply by rabbitism** [reply]
> 11 RC1 seems has not ItemsRepeater any more.

It's in an independent package.

**Reply by kyuranger** [reply]
I just installed it and it cannot be recognized as a control, reports error:
`Unable to resolve type ItemsRepeater from namespace https://github.com/avaloniaui Line 19, position 4.	AvaloniaApplication1	C:\Users\Admin\Desktop\AvaloniaApplication1\MainView.axaml	19	`
